docs/config_backend.md
# Konfiguration des Backends

Die Datei `./backend/volumes/config/env.example` enthält die Umgebungsvariablen, die für den Betrieb des Backends erforderlich sind. Diese Datei dient als Vorlage für die Konfigurationsdateien der verschiedenen Umgebungen (`env.dev`, `env.stage`, `env.prod`).

## Beispielkonfiguration (`env.example`)

```plaintext
# Allgemeine Konfiguration
CONFIG_NAME='Example Config'

# Logging
LOG_LEVEL=debug
LOG_DIR=./logs

# URLs & Ports
API_URL=http://localhost
API_PORT=3005
HOST_URL=http://localhost
HOST_PORT=8200

# Datenbank
DB_PATH=./volumes/database/hackathon.dev.db

# Secrets
JWT_SECRET=<your-secret-key>
API_SECRET=<your-api-secret>

# Sicherheit & Domains
ALLOWED_DOMAINS=your-domain.com,department-domain.com
ACTIVATION_URL=http://localhost:8200/account-activation

# SMTP-Konfiguration
SMTP_HOST=<SMTP_HOST>
SMTP_PORT=<SMTP_PORT>
SMTP_USER=<SMTP_USER>
SMTP_PASSWORD=<SMTP_PASSWORD>
SMTP_SECURE=true
SMTP_FROM=<SENDER_EMAIL>
SMTP_UNAUTHORIZED=false
```

## Beschreibung der Variablen

### Allgemeine Konfiguration
- **CONFIG_NAME**: Name der Konfiguration, dient zur Identifikation und besserem Verständniss der einzelnen Configurationen.

### Logging
- **LOG_LEVEL**: Log-Level (`debug`, `info`, `warn`, `error`).
- **LOG_DIR**: Verzeichnis, in dem die Log-Dateien gespeichert werden.

### URLs & Ports
- **API_URL**: Die Basis-URL der API, die vom Backend verwendet wird. Diese URL wird von anderen Diensten oder dem Frontend genutzt, um API-Anfragen zu stellen. Beispiel: http://localhost für lokale Entwicklung oder https://api.example.com für die Produktion.
- **HOST_PORT**: Der Port, auf dem die Anwendung erreichbar ist. Standardmäßig wird 8200 verwendet, kann aber je nach Umgebung angepasst werden. Beispiel: 80 für HTTP oder 443 für HTTPS.
- **HOST_URL**: Die URL des Hosts, auf dem die Anwendung läuft. Diese wird verwendet, um die Anwendung extern zugänglich zu machen. Beispiel: http://localhost für lokale Tests oder https://app.example.com für die Produktion.
- **API_PORT**: er Port, auf dem die API läuft. Dieser wird vom Backend verwendet, um Anfragen zu verarbeiten. Beispiel: 3005 für lokale Entwicklung.

### Datenbank
- **DB_PATH**: Der Pfad zur Datenbankdatei, die vom Backend verwendet wird. Diese Datei enthält alle Daten, die von der Anwendung gespeichert werden, wie Projekte, Teams und Benutzerinformationen. Beispiel: ./volumes/database/hackathon.dev.db für lokale Entwicklung.

### Secrets
- **JWT_SECRET**: Ein geheimer Schlüssel, der für die Erstellung und Validierung von JSON Web Tokens (JWT) verwendet wird. JWTs werden für die Authentifizierung und Autorisierung von Benutzern genutzt. Dieser Schlüssel sollte sicher und einzigartig sein. Beispiel: 'MeinSuperGeheimesPasswort'
- **API_SECRET**: Ein zusätzlicher geheimer Schlüssel, der für API-Zugriffe verwendet wird. Dieser wird zusätzlich zum persöhnlichen User Token genutzt, um sicherzustellen, dass nur autorisierte Dienste oder Benutzer auf bestimmte API-Endpunkte zugreifen können. Beispiel: 'MeinGeheimesApiPasswort'

### Sicherheit & Domains
- **ALLOWED_DOMAINS**: Eine Liste von Domains, die Zugriff auf die Anwendung haben dürfen. Benutzer mit einer Mail Adresse dieser Domain, die sich neu bei der Anwendung registrieren, werden nach erfolgreicher EMail aktivierung automatisch freigeschaltet. Alle anderen erhalten einen Gast Status und müssen erst von einem Organisator freigeschaltet werden. Beispiel: your-domain.com,department-domain.com..
- **ACTIVATION_URL**: Die URL, die für die Aktivierung von Benutzerkonten verwendet wird. Diese wird in E-Mails zur Kontoaktivierung eingebettet, damit Benutzer ihre Konten bestätigen können. Beispiel: http://localhost:8200/account-activation für lokale Tests oder https://app.example.com/account-activation für die Produktion.

### SMTP-Konfiguration
- **SMTP_HOST**: Hostname des SMTP-Servers.
- **SMTP_PORT**: Port des SMTP-Servers.
- **SMTP_USER**: Benutzername für den SMTP-Server.
- **SMTP_PASSWORD**: Passwort für den SMTP-Server.
- **SMTP_SECURE**: Gibt an, ob eine sichere Verbindung verwendet wird (`true` oder `false`).
- **SMTP_FROM**: Absenderadresse für E-Mails.
- **SMTP_UNAUTHORIZED**: Gibt an, ob unautorisierte SMTP-Verbindungen erlaubt sind (`true` oder `false`).

## Verwendung
1. Kopiere die Datei `env.example` in das Verzeichnis `./backend/volumes/config/` und benenne sie entsprechend der Umgebung (`env.dev`, `env.stage`, `env.prod`).
2. Passe die Werte in der Datei an deine spezifischen Anforderungen an.

## Hinweise
- Stelle sicher, dass sensible Informationen wie `JWT_SECRET`, `API_SECRET` und `SMTP_PASSWORD` nicht in öffentlichen Repositories geteilt werden. Alle Dateien in dem config Ordner (mit ausnahme der Example Datei) stehen daher auch auf der gitignore liste.
- Für den produktiven Betrieb sollten sichere und eindeutige Werte verwendet werden.

docs/config_frontend.md
# Konfiguration des Frontends

Die Datei `./frontend/volumes/config/.env.example` enthält die Umgebungsvariablen, die für den Betrieb des Frontends erforderlich sind. Diese Datei dient als Vorlage für die Konfigurationsdateien der verschiedenen Umgebungen (`env.dev`, `env.stage`, `env.prod`).

## Beispielkonfiguration (`.env.example`)

```plaintext
CONFIG_NAME='Develop Config'
APP_ID='hackathon-manager-frontend'

# Logging
LOG_LEVEL=debug

# URLs & Ports
VITE_API_URL=http://localhost
VITE_HOST_URL=http://localhost
VITE_HOST_PORT=8200
VITE_API_PORT=3005

# Support
SUPPORT_EMAIL='mail@your-helpdesk.com'

# Secrets
API_SECRET='your-api-secret'
```


## Beschreibung der Variablen

### Allgemeine Konfiguration
- **CONFIG_NAME**: Name der Konfiguration, dient zur Identifikation und besserem Verständniss der einzelnen Configurationen.
- **APP_ID**: Eine eindeutige Kennung für die Frontend-Anwendung, die zur Identifikation in Logs oder Debugging verwendet werden kann.

### Logging
- **LOG_LEVEL**: Log-Level (`debug`, `info`, `warn`, `error`).

### URLs & Ports
- **VITE_API_URL**: Die Basis-URL der API, die vom Backend verwendet wird. Diese URL wird von anderen Diensten oder dem Frontend genutzt, um API-Anfragen zu stellen. Beispiel: http://localhost für lokale Entwicklung oder https://api.example.com für die Produktion.
- **VITE_HOST_PORT**: Der Port, auf dem die Anwendung erreichbar ist. Standardmäßig wird 8200 verwendet, kann aber je nach Umgebung angepasst werden. Beispiel: 80 für HTTP oder 443 für HTTPS.
- **VITE_HOST_URL**: Die URL des Hosts, auf dem die Anwendung läuft. Diese wird verwendet, um die Anwendung extern zugänglich zu machen. Beispiel: http://localhost für lokale Tests oder https://app.example.com für die Produktion.
- **VITE_API_PORT**: er Port, auf dem die API läuft. Dieser wird vom Backend verwendet, um Anfragen zu verarbeiten. Beispiel: 3005 für lokale Entwicklung.

### Support
- **SUPPORT_EMAIL**: Die E-Mail-Adresse des Support-Teams, die in der Anwendung angezeigt wird, um Benutzern Hilfe anzubieten.

### Secrets
- **API_SECRET**: Ein zusätzlicher geheimer Schlüssel, der für API-Zugriffe verwendet wird. Dieser wird zusätzlich zum persöhnlichen User Token genutzt, um sicherzustellen, dass nur autorisierte Dienste oder Benutzer auf bestimmte API-Endpunkte zugreifen können. Dieses muss mit dem API_SECRET aus dem Backend übereinstimmtn. Beispiel: 'MeinGeheimesApiPasswort'


## Verwendung
1. Kopiere die Datei .env.example in das Verzeichnis ./frontend/volumes/config/ und benenne sie entsprechend der Umgebung (env.dev, env.stage, env.prod).
2. Passe die Werte in der Datei an deine spezifischen Anforderungen an.

## Hinweise
- Stelle sicher, dass sensible Informationen wie `JWT_SECRET`, `API_SECRET` und `SMTP_PASSWORD` nicht in öffentlichen Repositories geteilt werden. Alle Dateien in dem config Ordner (mit ausnahme der Example Datei) stehen daher auch auf der gitignore liste.
- Für den produktiven Betrieb sollten sichere und eindeutige Werte verwendet werden.

CHANGELOG.md
# 📦 Changelog

## Version 0.4.2
**Release Date:** 2025-10-09
### ✨ Neue Funktionen
- **Going Public**  
  - Das Project ist ab jetzt OpenSource
  
- **GitHub Project**  
  - Es ist auf einem öffentlichen Github Account für alle zugänglich
  - Die BuildPipelines sind auf GitHub umgezogen und als Actions implementiert
  - Es werden im Githup Reposetory Hub fertiger Docker Container bereitgestellt
- **Firmen eigene Installationen möglich**  
  - Eigene Versionen lassen sich über ein separates private Projekt ausrollen und mit lokaler Config ergänzen

### 🐞 Fehlerbehebungen & Verbesserungen

- **Kleine Verbesserungen**  
  - Config Anpassung für bessere Docker integration
  - Frontend ebenfalls vollständig auf .env Datei umgestellt
  - Projekt von Credentials bereinigt
  - Vorbereitung von OpenSource


## Version 0.3.3
**Release Date:** 2025-10-24

### 🐞 Fehlerbehebungen & Verbesserungen

- **Kleine Verbesserungen**  
  - Anzeige des Countdown bei laufenden Hackathon
  - Kopieren der Projektteilnehmer Mailadressen enthält nun bei Organisatoren alle Adresse
  - CSV Export der Projektliste ink. Teilnehmer

## Version 0.3.2
**Release Date:** 2025-10-23

### ✨ Neue Funktionen
- **Project**  
  - Neues Feld für Ort / Raum hinzugefügt
  - Projekt Detail zeigt den Raum an
  - Projekt Edit ermöglicht das bearbeiten des Raumes
  - Das Dashboard zeigt den Raum an

### 🐞 Fehlerbehebungen & Verbesserungen

- **Kleine Verbesserungen**  
  - Die Teamliste zeigt nur Teams die noch im Pitch oder Aktiv sind.
  - Das Dashboard berücksichtigt nun abgebrochende Projekte
  - Das Dashboard unterscheided registrierte Nutze und aktiv am Projekt beteiligte Teilnehmer


## Version 0.3.1
**Release Date:** 2025-10-22

### ✨ Neue Funktionen
- **Daten Update**  
  - Neue Projekte wurde hinzugefügt
  - Initiatoren wurden angepasst
- **Profile Seite** 
  - Die Profile Seite mit der Anzeige des eigenen Profils & der initiierten Projekten wurde hinzugefügt.
  - Das Ändern der eigenen Daten ist nun direkt über die Profilseite möglich.
  - Avatar Bilder hinzugefügt
  - **Passwort zurücksetzen für Organisatoren**  
    - Organisatoren können das Passwort eines Benutzers direkt über die Profilseite zurücksetzen.
    - Das neue Passwort wird zufällig generiert und per E-Mail an den Benutzer gesendet.
- **Teilnehmer Seite**  
  - Die Teilnehmer sind klickbar und es wird das jeweilige Profil geöffnet.
  - Organisatoren können dort auch die Profildaten direkt bearbeiten.
  - Organisatoren können auch die Rolle eines Profils verändern.
  - Legende mit Erläuterung des Farbschema hinzugefügt
- **Projekt Detail Seite**  
  - Organisatoren können Teilnehmer und Initiatoren vom Projekte entfernen.
  - Organisatoren können Nutzer als Teilnehmer o. Initiator zum Projekt hinzufügen.
  - Ein neues Teilnehmer Auswahl Modal unterstützt bei der Suche nach Teilnehmern.


### 🐞 Fehlerbehebungen & Verbesserungen

- **Verbesserte Darstellung der Buttons auf der Profilseite**  
  - Die Buttons "Anrufen" und "Kontakt teilen" auf der Profil Seite wurden optisch angepasst (runde Buttons, weniger prominent).
  - Die Buttons "Mail Adressen Export" und "MS Teams öffnen" auf der Projekt Detail Seite wurden optisch angepasst (runde Buttons, weniger prominent).
  - Hover-Effekte und einheitliches Design für die Buttons hinzugefügt.

## Version 0.3.0
**Release Date:** 2025-10-16

### ✨ Neue Funktionen

- **Stable Hackathon Release**
  - Feature completed MVP
- **Realtime Monitoring per Grafana**
  - Teamgröße pro Projekt
  - Anzahl der Projekte pro Event
  - Registrierte Nutzer insgesamt

### 🐞 Fehlerbehebungen & Verbesserungen
- Eventliste ist nun Klickbar und wechselt das aktuell angezeigte Event
- Status Badge auch auf der Projektliste (Teilnehmer, Initiator, Projekt abgebrochen)
- Teilnehmer & Inititatoren (bei mehr als einem) können nun direkt über das Frontend von Organisatoren gelöscht werden.


## Version 0.2.9
**Release Date:** 2025-10-16

### ✨ Neue Funktionen

- **Teilnehmer Liste**  
  - Farbliche Markierung von Gast & Orga Accounts
- **Backend**  
  - PW Reset Funktion implementiert
  - Absender Mail Adresse freischalten lassen um lange SPAM Filter Zeiten zu reduzieren

### 🐞 Fehlerbehebungen & Verbesserungen
- User konnten keine neuen Projekte anlegnen
- Initiatoren konnten ihre eigenen Projekte nicht editieren
- Erweitertes Prometheus Monitoring 


## Version 0.2.8
**Release Date:** 2025-10-15

### ✨ Neue Funktionen

- **Einstellbare Projekt Teamgröße**  
  - Anpassung der Projektliste mit  Anzeige der max. Teamgröße, basierend auf den individuellen Projekteinstellungen.
  - Anpassung der Projekt Detail Seite mit Anzeige der max. Teamgröße
  - Projekt Edit / Anlegen Seite ermöglicht das Bearbeiten der individuellen Teamgröße
  - Erweiterung der Backend APIs & Datenbank um die Projekt Teamgröße
- **MS Teams Link auf Projektseite**  
  - In der Kopfzeile des Projekts ist ein Button der direkt zum MS Teams Kanal weiterleitet.
  - Projekt Edit / Anlegen Seite ermöglicht das Bearbeiten der MS Teams Chanel ID
  - Erweiterung der Backend APIs & Datenbank um die MS Teams ID
- **Organisator Berechtigung**  
  - Auch Organisatoren dürfen ein Projekt bearbeiten
- **Mail Adressen Export**  
  - Organisatoren haben einen zusätzlichen Button in der Projekt Kopfzeile zum Exportieren aller Teilnehmer Mail Adressen.

### 🐞 Fehlerbehebungen & Verbesserungen
- **Filter auf Projektliste & Teilnehmer liste nicht einheitlich**  
  - Das Design der Elemente ist nun einheitlich
  - Beide Seiten filtern nun direkt beim tippen
  - Auch nach Namen der Initiatoren & Teilnehmern wird wieder gefiltert
- **Monitoring**  
  - Auslesen des Monitoring Backends per Prometheus
  - Aufsetzen eines Grafana Dashbord

## Version 0.2.7
**Release Date:** 2025-10-06

### ✨ Neue Funktionen
- **Neues Design**
  - Ein neues helleres Design das stärker an die Farben und Elemente 
  - verbesserter Kontrast

## Version 0.2.6
**Release Date:** 2025-09-06

### ✨ Neue Funktionen

backend/src/config.js
// be aware that backend is not using vite.env from frontend application, because backend run as seperate application process
// vite frontend variables are public so a security issue to don´t use it together in one .env!

const { log } = require('console');
const dotenv = require('dotenv');
const path = require('path');

const envMap = {
  production: 'prod',
  prod: 'prod',
  development: 'dev',
  dev: 'dev',
  staging: 'stage',
  stage: 'stage'
};

// Wähle das richtige .env-File basierend auf NODE_ENV
const envFile = '.env.' + (envMap[process.env.NODE_ENV?.toLowerCase()] || 'stage');

dotenv.config({ path: path.resolve(__dirname, `../volumes/config/${envFile}`) });

// Rufe die Logging-Funktion auf
logEnvironmentVariables();

module.exports = {
  name: require('../package.json').name,
  version: require('../package.json').version,
  node_env: envMap[process.env.NODE_ENV?.toLowerCase()] || 'dev',
  config_name: check(process.env.CONFIG_NAME, 'default'),
  logLevel: check(process.env.LOG_LEVEL, 'info'),
  apiUrl: check(process.env.API_URL, 'http://localhost'),
  hostUrl: check(process.env.HOST_URL, 'http://localhost'),
  hostPort: check(process.env.HOST_PORT, '8200'),
  apiPort: check(process.env.API_PORT, '3005'),
  dbPath: check(process.env.DB_PATH, './hackathon.db'),
  mysqlHost: check(process.env.MYSQL_HOST, 'localhost'),
  mysqlPort: check(process.env.MYSQL_PORT, 3306),
  mysqlUser: check(process.env.MYSQL_USER, 'root'),
  mysqlPassword: check(process.env.MYSQL_PASSWORD, 'password'),
  mysqlDb: check(process.env.MYSQL_DB, 'hackathon'),
  mysqlConnectionLimit: check(process.env.MYSQL_CONNECTION_LIMIT, 10),
  mysqlQueueLimit: check(process.env.MYSQL_QUEUE_LIMIT, 0),
  jwtSecret: check(process.env.JWT_SECRET, 'your-secret-key'),
  apiSecret: check(process.env.API_SECRET, 'your-api-secret'),
  allowedDomains: check(process.env.ALLOWED_DOMAINS, 'dummy.de'),
  activationUrl: check(process.env.ACTIVATION_URL, 'http://localhost:3000/api/user/activate'),
  smtpHost: check(process.env.SMTP_HOST, 'localhost'),
  smtpPort: check(process.env.SMTP_PORT, 587),
  smtpSecure: check(process.env.SMTP_SECURE, false),
  smtpUser: check(process.env.SMTP_USER, ''),
  smtpPassword: check(process.env.SMTP_PASSWORD, ''),
  smtpFrom: check(process.env.SMTP_FROM, ''),
  smtpRejectUnauthorized: check(process.env.SMTP_UNAUTOHORIZED, false),
  rateLimitWindowMs: check(process.env.RATE_LIMIT_WINDOW_MS, 15 * 60 * 1000),
  rateLimitMax: check(process.env.RATE_LIMIT_MAX, 100),
  check,
  logEnvironmentVariables
};

function check(value, defaultValue) {
  if (typeof value === 'undefined' || value === null) {
    return defaultValue;
  }
  if (typeof defaultValue === 'number') {
    return isNaN(Number(value)) ? defaultValue : Number(value);
  }
  if (typeof defaultValue === 'boolean') {
    return value === 'true' || value === true;
  }
  return value;
}

// Externe Logging-Funktion
function logEnvironmentVariables() {
  // Printout only in debug mode
  if (process.env.LOG_LEVEL !== 'debug') return;

  // Use console.log, because logger is not available at start time
  console.log('### Config Printout #########################################');
  console.log('Aktuelles Arbeitsverzeichnis:', process.cwd());
  console.log('Verzeichnis der Datei:', __dirname);

  console.log('CONFIG_NAME:', process.env.CONFIG_NAME);
  console.log('LOG_LEVEL:', process.env.LOG_LEVEL);
  console.log('API_URL:', process.env.API_URL);
  console.log('API_PORT:', process.env.API_PORT);
  console.log('HOST_URL:', process.env.HOST_URL);
  console.log('HOST_PORT:', process.env.HOST_PORT);
  console.log('SMTP_HOST:', process.env.SMTP_HOST);
  console.log('SMTP_PORT:', process.env.SMTP_PORT);
  console.log('SMTP_SECURE:', process.env.SMTP_SECURE);
  console.log('SMTP_USER:', process.env.SMTP_USER);

  console.log('Aktueller Modus (NODE_ENV):', process.env.NODE_ENV || 'nicht gesetzt');
  console.log('Geladene .env-Datei:', envFile);
  console.log('#############################################################');
}
